third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/MemoryOpToLLVM.cpp

```diff
@@ -136,64 +144,143 @@ struct LocalLoadOpConversion
+  if (!maybeQuot.has_value()) {
     return failure();
   }
+  auto quot = maybeQuot.value();
 
-  auto *ctx = rewriter.getContext();
+  // Compute the registers that don't broadcast to see how many elements we can
+  // vectorise over
+  auto mask =
+      (~quot.getFreeVariableMasks()[kReg]) & (quot.getInDimSize(kReg) - 1);
+  auto vec = std::min(2, __builtin_popcount(mask));
+  // 2) NYI. stmatrix.x4 is the only supported for now
+  if (vec != 2) {
+    return failure();
+  }
 
-  auto layout =
-      chooseStMatrixLayout(rewriter.getContext(), srcTy, swizzleByteSize);
-  auto llvmElemTy = typeConverter->convertType(memDescType.getElementType());
-  auto smemPtrTy = ptr_ty(ctx, 3);
+  uint32_t newMask = 0;
+  for (int i = 0; i < vec; ++i) {
+    auto lowest = mask & -mask;
+    newMask |= lowest;
+    mask -= lowest;
+  }
+  mask = newMask;
 
-  auto kRegister = str_attr("register");
-  auto kLane = str_attr("lane");
-  auto kWarp = str_attr("warp");
-  auto kBlock = str_attr("block");
+  auto srcVals = unpackLLElements(loc, adaptorSrc, rewriter);
+  // If the permutation is non-trivial, we need to permute the layout and values
+  if (!llvm::isPowerOf2_32(mask + 1)) {
+    // 1) Move the bases that we want to use to the lowest bits
+    quot = permuteInDimToFront(quot, kReg, mask);
+    // 2) Permute the values accordingly
+    // We don't need to touch the values that go in each segement
+    auto n = tile.getInDimSizeLog2(kReg);
+    auto maskReps = (mask << n) | ((1 << n) - 1);
+    LinearLayout permRegs = permuteInDimToFront(
+        LinearLayout::identity1D(srcVals.size(), kReg, kReg), kReg, maskReps);
+    SmallVector<Value> permVals(srcVals.size());
+    for (int i = 0; i < srcVals.size(); i++) {
+      auto srcIdx = permRegs.apply({{kReg, i}}).begin()->second;
+      permVals[i] = srcVals[srcIdx];
+    }
+    srcVals = std::move(permVals);
+  }
+  auto reps = zerosLike(tile) * quot;
 
+  // Compute the common components of the register
   auto [laneId, warpId] = getLaneAndWarpId(rewriter, loc);
 
-  auto regBase = applyLinearLayout(loc, rewriter, layout,
-                                   {{kRegister, b.i32_val(0)},
-                                    {kLane, laneId},
-                                    {kWarp, warpId},
-                                    {kBlock, b.i32_val(0)}})[0]
+  // FIXME(Lezcano): Should we bail if any of the other 3 lane bases is zero?
+
+  // Just implement .x4 for now
+  // We want to use the last 2 bits of the lane id to select the start of the
+  // vectorisation and the fist 3 bits to select the column
+
+  // FIXME Change the constants below / pack things into uint32_t to support
+  // bitwidths other than 16
+  assert(bitwidth == 16);
+  auto regBase = applyLinearLayout(loc, rewriter, quot,
+                                   {{kReg, b.lshr(laneId, b.i32_val(3))},
+                                    {kLane, b.and_(laneId, b.i32_val(0x7))},
+                                    {kWarp, warpId}})[0]
```

Comment:
What's going on here, seems like some abuse is going on?

Reply:
abuse? This is just implementing the address map that ldmatrix asks for:
![image](https://github.com/user-attachments/assets/4f1e6434-a4b3-4e4a-a4a6-51acc6a09840)
In particular, the lower three bits of the lane should map to the columns, while the top two bits should map to the 4 different matrices (given by the first 2 basis of the reps of quot). I'll write a comment.

Reply:
But you're setting the register index based on the lane id so clearly you're abusing the labels `kReg` and `kLane to mean something else. That's pretty confusing.

Reply:
I'm a bit concerned that this new feature that is supposed to simplify all our lowerings is resulting in (for me) unreadable code.

Reply:
Okay, I understand now. When using `.x4` you pass the pointer corresponding to the row for register `T0:r[i]` in `T(0 + 8*i):p` so the lane really is giving the register index. Makes sense. I think I would restate your comment though as:
```
  // Here we implement the stmatrix.x4 addressing. In particular, the row pointers
  // for each submatrix r in thread t are communicated by the stmatrix call in 
  // laneId = (t // 8) + (8 * r), so r = laneId // 8 and t = laneId % 8.
```

Reply:
I rewrote it in a third way, tell me WDYT.